packages/ts/react-grid/src/autogrid.tsx

```diff
@@ -0,0 +1,97 @@
+import type { ModelConstructor } from '@hilla/form';
+import {
+  Grid,
+  GridProps,
+  type GridDataProviderCallback,
+  type GridDataProviderParams,
+  type GridElement,
+} from '@hilla/react-components/Grid.js';
+import { GridSortColumn } from '@hilla/react-components/GridSortColumn.js';
+import { useEffect, useRef } from 'react';
+import type { CrudService } from './crud';
+import { getProperties } from './modelutil.js';
+
+// import Sort from "Frontend/generated/dev/hilla/mappedtypes/Sort";
+// import Direction from "Frontend/generated/org/springframework/data/domain/Sort/Direction";
+type Sort = any;
+enum Direction {
+  ASC = 'ASC',
+  DESC = 'DESC',
+}
+
+export type _AutoGridProps<TItem> = {
+  service: CrudService<TItem>;
+  model: ModelConstructor<TItem, any>;
+};
+export type AutoGridProps<TItem> = GridProps<TItem> & _AutoGridProps<TItem>;
+
+const createDataProvider = <TItem,>(grid: GridElement<TItem>, service: CrudService<TItem>) => {
+  const listMethod = service.list;
+  let first = true;
+
+  return async (params: GridDataProviderParams<TItem>, callback: GridDataProviderCallback<TItem>) => {
+    const sort: Sort = {
+      orders: params.sortOrders.map((order) => ({
+        property: order.path,
+        direction: order.direction == 'asc' ? Direction.ASC : Direction.DESC,
+        ignoreCase: false,
+      })),
+    };
+
+    const pageNumber = params.page;
+    const pageSize = params.pageSize;
+    const req = {
+      pageNumber,
+      pageSize,
+      sort,
+    };
+
+    const items = await listMethod(req);
+    let size;
+    if (items.length === pageSize) {
+      size = (pageNumber + 1) * pageSize + 1;
+      if (size < (grid as any)._cache.size) {
+        // Only allow size to grow here to avoid shrinking the size when scrolled down and sorting
+        size = undefined;
+      }
+    } else {
+      size = pageNumber * pageSize + items.length;
+    }
```

Comment:
We could also return the size from the crud service and then just use it here. Related to that, the current signature of `CrudService` would prevent use from ever doing so as it only returns a list.

How about something like this:
```java
public class PageResult<T> {
    private final List<T> content;
    private final long totalElements;

    public PageResult(Page<T> page) {
        this.content = page.getContent();
        this.totalElements = page.getTotalElements();
    }
}

public interface CrudService<T> {
    PageResult<T> list(Pageable pageable);
}

@EndpointExposed
public class CrudRepositoryService<T, ID> implements CrudService<T> {
    ...
    @Override
    public PageResult<T> list(Pageable pageable) {
        return new PageResult(repository.findAll(pageable));
    }
}
```

Reply:
Yes this is done intentionally for now so you do not have to worry about the performance hit of making an extra query and the more complex return type. If you return `PageResult` then the frontend code gets more complex. What is in autogrid does not matter but the crud services will be used by people outside autogrid also. 

The `Page` class in Spring Boot actually can contain the total elements so it is kind of your `PageResult` already. Hilla just ignores that #703

Reply:
What would be your long-term plan for this? A second method for counting that you either implement or not? IMO this hurts usability as you can not scroll to the end of the grid, you have to go through all pages first.

Reply:
Should we leave this as an additional feature, to optionally include the size? The current one only implements infinite scrolling

Reply:
Whether not being able to scroll directly to the end or not is a problem is quite use case dependent. Some people search instead of scroll, other people don't. We must have another method for the case as it has different requirements

Reply:
Let's cover this later with a separate API then.